### PR TITLE
federation: Adding support for namespace admission controls in federation-apiserver

### PR DESCRIPTION
**Pull request by nikhiljindal**
Now that we have namespaces in federation apiserver, we can support namespace admission controls.

There are 3 of these:
namespace/autoprovision, namespace/exists and namespace/lifecycle.
namespace/autoprovision, namespace/exists should be deprecated in kubernetes(https://github.com/kubernetes/kubernetes/issues/31195). Adding support for namespace/lifecycle to federation-apiserver.
As in kube-apiserver, enabling namespace/lifecycle by default.

``` release-note
Action required: If you have a running federation control plane, you will have to ensure that for all federation resources, the corresponding namespace exists in federation control plane.

federation-apiserver now supports NamespaceLifecycle admission control, which is enabled by default. Set the --admission-control flag on the server to change that.
```

cc @kubernetes/sig-cluster-federation @quinton-hoole

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31139)

### CODE REVIEW

**Comment by nikhiljindal** [comment]
Am working on fixing the federation e2es.

**Comment by nikhiljindal** [comment]
cc @derekwaynecarr

**Comment by nikhiljindal** [comment]
This is now ready for review. Have fixed all federation e2es and have added a test to ensure that a request to create a namespaced resource fails if the namespace does not exist.

I have a TODO to clean up the federation namespace after each test, which I will send another PR for (need to fix https://github.com/kubernetes/kubernetes/issues/31077 before that).

ptal.

**Comment by derekwaynecarr** [comment]
I think you can only add NamespaceLifecycle.  The others should be marked
deprecated.

**Comment by nikhiljindal** [comment]
Removed the other 2 namespace admission controls and filed https://github.com/kubernetes/kubernetes/issues/31195 to keep track to deprecate it in kube-apiserver.

File: federation/cluster/common.sh [hunk 1]
```diff
@@ -94,7 +94,7 @@ function create-federation-api-objects {
     export FEDERATION_API_NODEPORT=32111
     export FEDERATION_NAMESPACE
     export FEDERATION_NAME="${FEDERATION_NAME:-federation}"
-    export DNS_ZONE_NAME="${DNS_ZONE_NAME:-federation.example}"  # See https://tools.ietf.org/html/rfc2606
+    export DNS_ZONE_NAME="${DNS_ZONE_NAME:-federation.example.}"  # See https://tools.ietf.org/html/rfc2606
```
Comment (quinton-hoole) [inline federation/cluster/common.sh]: Doesn't seem to belong in this PR?

File: test/e2e/federation-apiserver.go [hunk 2]
```diff
@@ -84,8 +84,44 @@ var _ = framework.KubeDescribe("Federation apiserver [Feature:Federation]", func
 			framework.Logf("Verified that zero clusters remain")
 		})
 	})
+	Describe("Admission control", func() {
+		AfterEach(func() {
+			framework.SkipUnlessFederated(f.Client)
```
Comment (quinton-hoole) [inline test/e2e/federation-apiserver.go]: Surely this belongs in BeforeEach(), not AfterEach()?

File: test/e2e/federation-apiserver.go [hunk 3]
```diff
@@ -84,8 +84,44 @@ var _ = framework.KubeDescribe("Federation apiserver [Feature:Federation]", func
...
+		})
+
+		It("should not be able to create resources if namespace does not exist", func() {
+			framework.SkipUnlessFederated(f.Client)
```
Comment (quinton-hoole) [inline test/e2e/federation-apiserver.go]: This belongs in BeforeEach()?

File: test/e2e/federation-apiserver.go [hunk 4]
```diff
@@ -84,8 +84,44 @@ var _ = framework.KubeDescribe("Federation apiserver [Feature:Federation]", func
...
+
+			// Creating a service in a non-existing namespace should fail.
+			svcNamespace := "federation-admission-test-ns"
```
Comment (quinton-hoole) [inline test/e2e/federation-apiserver.go]: ideally use a unique namespace name (as generated by the framework), so that running the test multiple times against the same cluster (e.g. as the soak tests do), cannot cause contamination between runs.

File: test/e2e/federation-apiserver.go (hunk 3 above)
Comment (nikhiljindal) [inline test/e2e/federation-apiserver.go]: Sorry did not get you. Do you mean move SkipUnlessFederated to BeforeEach?
We put that line in BeforeEach, AfterEach and It() - all 3.

File: test/e2e/federation-util.go [hunk 5]
```diff
@@ -281,7 +289,7 @@ func deleteServiceOrFail(clientset *federation_release_1_4.Clientset, namespace
 	framework.ExpectNoError(err, "Error deleting service %q from namespace %q", serviceName, namespace)
 }
 
-func podExitCodeDetector(f *framework.Framework, name string, code int32) func() error {
+func podExitCodeDetector(f *framework.Framework, name, namespace string, code int32) func() error {
```
Comment (quinton-hoole) [inline test/e2e/federation-util.go]: nit: you could use types.NamespacedName here instead.

**Comment by quinton-hoole** [comment]
Thanks @nikhiljindal. Comments can be addressed in a followup PR.  LGTM.

**Comment by quinton-hoole** [comment]
@nikhiljindal One of these tests is ~perma-failing since 8/23.  Could you please take a look?

https://k8s-testgrid.appspot.com/google-gce#gce-federation

**Comment by nikhiljindal** [comment]
Sure. Am on it
